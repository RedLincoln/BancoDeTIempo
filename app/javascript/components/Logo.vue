<template>
  <v-img :max-width="maxWidth" :src="src"> </v-img>
</template>

<script>
import logoDark from "images/logo_dark.png";
import logo from "images/logo.png";

export default {
  props: {
    maxWidth: {
      type: String,
      default: "150",
    },
    dark: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    src() {
      return this.dark ? logoDark : logo;
    },
  },
};
</script>

<style>
</style>