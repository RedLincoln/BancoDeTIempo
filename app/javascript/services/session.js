import { axiosInstance as axios } from "./axios";
import { setToken } from "./jwt";

export default {
  authenticate(authData) {
    return axios
      .post("/api/session.json", authData)
      .then((response) => {
        const { user_name, token } = response.data;
        setToken(token);
        return user_name;
      })
      .catch((err) => {
        return err.response;
      });
  },
  register(formData) {
    return axios.post("/api/sign_up.json", formData);
  },
  logOut() {},
};
