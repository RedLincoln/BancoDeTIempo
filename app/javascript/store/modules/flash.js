export const namespaced = true;

export const state = {
  flashs: [],
};

export const mutations = {
  APPEND_FLASH(state, flash) {
    state.flashs.push(flash);
  },
  REMOVE_FIRST_FLASH(state) {
    state.flashs.shift();
  },
  RESET_FLASH(state) {
    state.flashs = [];
  },
};

export const actions = {
  removeAllFlash({ commit }) {
    commit("RESET_FLASH");
  },
  addFlash({ commit }, flash) {
    commit("APPEND_FLASH", flash);
  },
};
