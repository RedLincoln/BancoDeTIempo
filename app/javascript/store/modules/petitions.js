import PetitionService from "../../services/petitions";

const { getPetitions } = PetitionService;

export const namespaced = true;

export const state = {
  petitions: [],
  total: 0,
};

export const mutations = {
  SET_PETITIONS(state, petitions) {
    state.petitions = petitions;
  },
  SET_TOTAL(state, total) {
    state.total = total;
  },
};

export const actions = {
  fetch({ commit, dispatch }, { q }) {
    return dispatch("setLoading", true, { root: true }).then(() => {
      getPetitions({ q }).then((data) => {
        commit("SET_PETITIONS", data.petitions);
        commit("SET_TOTAL", data.total);
        dispatch("setLoading", false, { root: true });
      });
    });
  },
};

export const getters = {
  all(state) {
    return state.petitions;
  },
  clients: (state) => (user_id) => {
    return state.petitions.filter((petition) => petition.client.id === user_id);
  },
  owners: (state) => (user_id) => {
    return state.petitions.filter((petition) => petition.client.id !== user_id);
  },
};
