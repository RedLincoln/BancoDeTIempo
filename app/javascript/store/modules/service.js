import ServiceService from "../../services/services";

const { getCategories, getTags } = ServiceService;

export const namespaced = true;

export const state = {
  tags: [],
  categories: [],
};

export const mutations = {
  SET_TAGS(state, tags) {
    state.tags = tags;
  },
  SET_CATEGORIES(state, categories) {
    state.categories = categories;
  },
};

export const actions = {
  fetchTagsAndCategories({ commit }) {
    return Promise.all([getCategories(), getTags()]).then(
      ([categories, tags]) => {
        commit("SET_TAGS", tags);
        commit("SET_CATEGORIES", categories);
        return;
      }
    );
  },
};
