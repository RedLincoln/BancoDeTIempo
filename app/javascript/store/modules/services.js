import ServicesService from "../../services/services";

const { getServices, getCategories } = ServicesService;

export const namespaced = true;

export const state = {
  services: [],
  perPage: 10,
  totalServices: 0,
  categories: [],
  tags: [],
};

export const mutations = {
  SET_SERVICES(state, services) {
    state.services = services;
  },
  SET_TOTAL(state, total) {
    state.totalServices = total;
  },
  SET_CATEGORIES(state, categories) {
    state.categories = categories;
  },
};

export const actions = {
  fetch(
    { commit, state, dispatch },
    { page, q, service_type, sort_order, category }
  ) {
    dispatch("setLoading", true, { root: true }).then(() => {
      getServices({
        limit: state.perPage,
        offset: page * state.perPage,
        q,
        service_type,
        sort_order,
        category,
      }).then((response) => {
        commit("SET_SERVICES", response.data.services);
        commit("SET_TOTAL", response.data.total);
        dispatch("setLoading", false, { root: true });
      });
    });
  },
  fetchCategories({ commit, dispatch }) {
    return dispatch("setLoading", true, { root: true }).then(() => {
      getCategories().then((categories) => {
        commit("SET_CATEGORIES", categories);
        dispatch("setLoading", false, { root: true });
      });
    });
  },
};
