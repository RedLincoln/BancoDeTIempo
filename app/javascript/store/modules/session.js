import SessionService from "../../services/session";

export const namespaced = true;

const { authenticate, register } = SessionService;

export const state = {};

export const mutations = {};

export const actions = {
  logIn({}, authInfo) {
    return authenticate(authInfo);
  },
  signUp({}, formData) {
    return register(formData);
  },
};

export const getters = {};
