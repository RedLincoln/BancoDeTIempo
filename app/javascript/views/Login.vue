<template>
  <div>
    <v-alert
      data-testid="error-message"
      v-show="errorMessage.length > 0"
      border="right"
      colored-border
      type="error"
      elevation="2"
    >
      {{ errorMessage }}
    </v-alert>
    <v-form data-testid="login-form">
      <v-text-field v-model="email" data-testid="email-field"> </v-text-field>
      <v-text-field
        v-model="password"
        data-testid="password-field"
      ></v-text-field>
      <v-btn @click="logIn" data-testid="submit-btn"></v-btn>
    </v-form>
  </div>
</template>

<script>
export default {
  data() {
    return {
      email: "",
      password: "",
      errorMessage: "",
    };
  },
  methods: {
    logIn() {
      this.$store
        .dispatch("session/logIn", {
          email: this.email,
          password: this.password,
        })
        .then((number) => {
          this.$router.push({ name: "home" });
        })
        .catch(({ message }) => {
          this.errorMessage = message;
        });
    },
  },
};
</script>

<style>
</style>