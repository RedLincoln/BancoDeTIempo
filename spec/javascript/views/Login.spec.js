import Vuetify from "vuetify";
import Vuex from "vuex";

import { createLocalVue, mount } from "@vue/test-utils";
import Login from "../../../app/javascript/views/Login.vue";
import Logo from "../../../app/javascript/components/Logo.vue";
import { actions } from "../../../app/javascript/store/modules/session";

const { logIn } = actions;

describe("Vuex have needed actions", () => {
  it("logIn", () => {
    expect(logIn).not.toBeUndefined();
  });
});

describe("Login.vue", () => {
  const localVue = createLocalVue();
  localVue.component("Logo", Logo);
  let vuetify;
  let wrapper;
  let actions;
  let store;

  let $router;

  const userInfo = {
    email: "michael@example.com",
    password: "password",
  };

  beforeEach(() => {
    jest.resetAllMocks();

    $router = {
      push: jest.fn(),
    };

    localVue.use(Vuex);

    actions = {
      logIn: jest.fn(),
    };
    store = new Vuex.Store({
      modules: {
        session: {
          namespaced: true,
          actions,
        },
      },
    });

    vuetify = new Vuetify();
    wrapper = mount(Login, {
      localVue,
      vuetify,
      store,
      mocks: {
        $router,
      },
    });
  });

  it("renders properly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("have the necesary elements", () => {
    expect(wrapper.find('[data-testid="login-form"]').exists()).toBe(true);
    expect(wrapper.find('[data-testid="email-field"]').exists()).toBe(true);
    expect(wrapper.find('[data-testid="password-field"]').exists()).toBe(true);
    expect(wrapper.find('[data-testid="submit-btn"]').exists()).toBe(true);
  });

  it("successfuly logIn", async () => {
    actions.logIn.mockResolvedValue();
    await wrapper.find('[data-testid="email-field"]').setValue(userInfo.email);
    await wrapper
      .find('[data-testid="password-field"]')
      .setValue(userInfo.password);
    await wrapper.find('[data-testid="submit-btn"]').trigger("click");

    expect(actions.logIn).toHaveBeenCalledTimes(1);
    expect(actions.logIn).toHaveBeenCalledWith(
      expect.objectContaining({}),
      expect.objectContaining(userInfo)
    );

    await wrapper.vm.$nextTick();
    expect($router.push).toHaveBeenCalledWith({ name: "home" });
  });

  it("bad logIn", async () => {
    const errorMessage = "Credentials problem";
    actions.logIn.mockRejectedValue({
      message: errorMessage,
    });

    await wrapper.find('[data-testid="email-field"]').setValue(userInfo.email);
    await wrapper
      .find('[data-testid="password-field"]')
      .setValue(userInfo.password);
    await wrapper.find('[data-testid="submit-btn"]').trigger("click");

    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();

    expect(wrapper.find('[data-testid="error-message"]').text()).toBe(
      errorMessage
    );
  });
});
