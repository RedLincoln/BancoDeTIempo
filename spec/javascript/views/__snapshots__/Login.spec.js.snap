// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue renders properly 1`] = `
<div>
  <div role="alert" class="v-alert v-sheet theme--light elevation-2 v-alert--border v-alert--border-right" data-testid="error-message" style="display: none;">
    <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--light error--text"></i>
      <div class="v-alert__content">

      </div>
      <div class="v-alert__border v-alert__border--right error v-alert__border--has-color"></div>
    </div>
  </div>
  <form novalidate="novalidate" data-testid="login-form" class="v-form">
    <div class="v-input theme--light v-text-field">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><input data-testid="email-field" id="input-7" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <div class="v-input theme--light v-text-field">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><input data-testid="password-field" id="input-10" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div> <button type="button" class="v-btn v-btn--contained theme--light v-size--default" data-testid="submit-btn"><span class="v-btn__content"></span></button>
  </form>
</div>
`;
